text/0255-sensitive-data-redaction.md

```diff
@@ -0,0 +1,397 @@
+        redaction: 's/^[^@]*@/REDACTED@/'
+  ```
+
+The above is just an example for how such a configuration may look like, different
+formats are discussed below.
+
+#### Redaction requirement profiles
+
+An end user should be able to pick from a set of redaction requirement profiles
+that cover common configurations:
+
+- `NONE` or `NEVER`: No redaction is applied. This is useful if an end user
+  chooses to run a collector nearby the application that will take care of
+  redaction out of process. If this profile is selected the SDK must make sure
+  that (almost) no overhead remains from the redaction methods.
+- `DEFAULT`: A set of rules that especially make sure that no security-sensitive
+  data is exposed, only limited PII-related redactions are applied.
+- `STRICT`: A set that builds on top of `DEFAULT` and also applies commonly required
+  PII-related redactions (anonymize emails, names, IPs, etc)
+  `STRICTER`: A set that builds on top of `STRICT` and applies even more redactions,
+  e.g. drop all query attribute values, drop PII-related data, etc.
+- `DROP` or `ALWAYS`: All values with (potentially) sensitivity details will be dropped
+
+Those profiles will be defined through the semantic conventions (see below).
+
+#### Fine-grained configuration
+
+On top of the redaction requirements profiles an end user should have capabilities
+to add their own local requirements, either through configuration or through code.
+
+The configuration is a list of rules that the SDK can parse and apply. An item
+in this list will consist of:
+
+- Conditions when a redaction should be applied
+- Instructions for the redaction that will be applied when the conditions are met
+
+In an example configuration file a list could look like the following:
+
+```yaml
+  sensitiveData:
+      - attribute: url.query
+        redaction: REDACT_ALL_URL_VALUES
+      - attribute: com.example.customer.name
+        redaction: DROP
+      - attribute: com.example.customer.email
+        redaction: 's/^[^@]*@/REDACTED@/'
+```
+
+In this example the condition is expressed via `attribute` and the instructions
+via `redaction`, where either a constant is selected that runs a predefined helper
+method, or a sed-like expression that uses a regular expression to apply redaction.
+
+Another possibility is to re-use a language like
+[OTTL](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/README.md).
+
+### Semantic Conventions
+
+To enable the redaction requirement profiles the semantic conventions need to be
+annotated with sensitivity details.
+
+Example:
+
+| Attribute | ... existing columns ... | sensitivity details |
+|-----------|--------------------------|---------------------|
+| `url.query`|                         | Rationale: Some verbatim wording why this is the way it is below<br>Type: `mixed`<br>`DEFAULT`: `REDACT_INSECURE_URL_PARAM_VALUES`<br>STRICT: `REDACT_ALL_URL_VALUES`<br>`STRICTER`: `DROP` |
```

Comment:
Can you have this proposal be against semconv YAML model (or the model in general) vs. the documentation?

The model is what we'll be generating code against and what we'll be advertising via SchemaURL.  It'd be a lot easier to comment/design around it.

Reply:
Yes, I can do that!

Reply:
@lquerel my knowledge about telemetry schema is very limited, can you provide me some examples and pointers to learn more about this? Overall I am a fan of making one thing a "subset of another", but I would need to understand the impact of that, technically but also organizational (e.g. what would be the extended timeline for that?)